Framework/Algorithms/src/PolarizationCorrections/FlipperEfficiency.cpp

```diff
@@ -0,0 +1,126 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2024 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#include <filesystem>
+
+#include "MantidAPI/Axis.h"
+#include "MantidAPI/FileProperty.h"
+#include "MantidAPI/WorkspaceGroup.h"
+#include "MantidAlgorithms/PolarizationCorrections/FlipperEfficiency.h"
+#include "MantidAlgorithms/PolarizationCorrections/PolarizationCorrectionsHelpers.h"
+#include "MantidAlgorithms/PolarizationCorrections/SpinStateValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/Unit.h"
+
+namespace {
+/// Property Names
+namespace PropNames {
+auto constexpr INPUT_WS{"InputWorkspace"};
+auto constexpr OUTPUT_WS{"OutputWorkspace"};
+auto constexpr OUTPUT_FILE{"OutputFilePath"};
+auto constexpr SPIN_STATES{"SpinStates"};
+} // namespace PropNames
+
+auto constexpr FILE_EXTENSION{".nxs"};
+auto constexpr INITIAL_SPIN{"11,10,01,00"};
+} // namespace
+
+namespace Mantid::Algorithms {
+
+using namespace API;
+using namespace Kernel;
+using PolarizationCorrectionsHelpers::workspaceForSpinState;
+
+// Register the algorithm in the AlgorithmFactory
+DECLARE_ALGORITHM(FlipperEfficiency)
+
+std::string const FlipperEfficiency::summary() const { return "Calculate the efficiency of the polarization flipper."; }
+
+void FlipperEfficiency::init() {
+  declareProperty(std::make_unique<WorkspaceProperty<WorkspaceGroup>>(PropNames::INPUT_WS, "", Direction::Input),
+                  "Group workspace containing flipper transmissions for all 4 polarisation states.");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>(PropNames::OUTPUT_WS, "", Direction::Output,
+                                                                       PropertyMode::Optional),
+                  "Workspace containing the wavelength-dependent efficiency for the flipper.");
+  declareProperty(std::make_unique<FileProperty>(PropNames::OUTPUT_FILE, "", FileProperty::OptionalSave),
+                  "File name or path for the output to be saved to.");
+  auto const spinValidator = std::make_shared<SpinStateValidator>(std::unordered_set<int>{4});
+  declareProperty(PropNames::SPIN_STATES, INITIAL_SPIN, spinValidator,
+                  "Order of individual spin states in the input group workspace, e.g. \"01,11,00,10\"");
+}
+
+std::map<std::string, std::string> FlipperEfficiency::validateInputs() {
```

Comment:
Should the input workspaces contain a single spectrum? I had a go at passing in a Zoom workspace group converted into wavelength but without summing to a single spectrum. The algorithm still output an efficiency workspace but with multiple spectra. I think this may not necessarily be wrong, but only one spectrum of efficiencies would be relevant to eventually pass into the correction algorithm. If we're not sure, this is something we could make a note of and discuss with the users as part of their testing potentially.

Reply:
I think we're not certain on this one. I didn't want to restrict it too much on the grounds that there might be more than one monitor they want to look at. It's something I'll include in the email when informing them it's available for testing.